core/src/main/java/com/oath/oak/ValueUtilsImpl.java

```diff
@@ -62,44 +62,53 @@ public void unsafeBufferToIntArrayCopy(ByteBuffer bb, int srcPosition, int[] dst
     }
 
     @Override
-    public <V> ValueResult put(Chunk<?, V> chunk, Chunk.LookUp lookUp, V newVal, OakSerializer<V> serializer,
-                               MemoryManager memoryManager) {
+    public <V> ValueResult put(Chunk<?, V> chunk, EntrySet.LookUp lookUp, V newVal, OakSerializer<V> serializer,
+        MemoryManager memoryManager, InternalOakMap internalOakMap) {
         ValueResult result = lockWrite(lookUp.valueSlice, lookUp.version);
         if (result != TRUE) {
             return result;
         }
-        Slice s = innerPut(chunk, lookUp, newVal, serializer, memoryManager);
+        Slice s = lookUp.valueSlice;
+        int capacity = serializer.calculateSize(newVal);
+        if (capacity + getHeaderSize() > s.getByteBuffer().remaining()) {
+            s = moveValue(chunk, lookUp, memoryManager, internalOakMap, newVal);
+            if (s == null) {
+                // rebalance was needed or the entry was updated by someone else, need to retry
+                unlockWrite(lookUp.valueSlice);
+                return RETRY;
+            }
+        } else {
+            s = innerPut(chunk, lookUp, newVal, serializer, memoryManager);
+        }
         unlockWrite(s);
         return TRUE;
```

Comment:
It is a bit more complicated as we can not always unlock the "old" slice. We might get a new slice in the middle. I have added a comments. I see that this code bothers you a lot, if you want we can speak about it. I wrote it this way also because of delicate synchronization issues involved in moving the value...